File: Exercises/group_anagrams.py
# Topic: Group Anagrams

# Given an array of strings strs, group the anagrams together. You can return the answer in any order.
# An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.
# Example 0:
# Input: strs = ["eat","tea","tan","ate","nat","bat"]
# Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
# Example 1:
# Input: strs = [""]
# Output: [[""]]
# Example 2:
# Input:  strs = ["a"]
# Output: [["a"]]

# Constraints
# 1 <= strs.length <= 104
# 0 <= strs[i].length <= 100
# strs[i] consists of lowercase English letters.
def is_anagram(output_array_str, str):
  #  if len != return false
  output_array_str_sorted = sorted(output_array_str)
  str_sorted = sorted(str)
  if (output_array_str_sorted == str_sorted):
    return True
  return False

def group_anagrams(string_array):
  #  check empty
  if len(string_array) == 0:
    return []
  output_array = []
  for element in string_array:
    add_flag = 0
    for sub_list in output_array:
      if(is_anagram(sub_list[0], element)):
        sub_list.append(element)
        add_flag = 1
    if add_flag == 0:
      new_sub_list = [element]
      output_array.append(new_sub_list)
    # print(output_array)

  return output_array

strs = ["eat","tea","tan","ate","nat","bat"]
strs2 = [""]
strs3 = ["a"]
print(group_anagrams(strs))
print(group_anagrams(strs2))
print(group_anagrams(strs3))

File: Exercises/GroupAnagrams.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Map;
import java.util.HashMap;

//Time Complexity: O(NKlogK), where N is the length of strs, and K is the max length of a string in
//strs. The outer loop is O(N) because we're iterating through each string, and sorting each string
//is O(KlogK) time.
//Space Complexity: O(NK), the total amount of information stored in output

class GroupAnagrams {                                                                               
  public static void main(String[] args) {                                                          
    String[] strs1 = {"eat","tea","tan","ate","nat","bat"};                                         
    String[] strs2 = {""};                                                                          
    String[] strs3 = {"a"};                                                                         
    System.out.println(groupAnagrams(strs1));                                                       
    System.out.println(groupAnagrams(strs2));                                                       
    System.out.println(groupAnagrams(strs3));                                                       
  }                                                                                                 

  //Maintain a map where each key is a sorted string and its values are a list of words in str
  //that are anagrams of the key                                                                    
  public static List<List<String>> groupAnagrams(String[] strs) {
    List<List<String>> output = new ArrayList<>();
    if (strs.length == 0) {
      return output;
    }
    Map<String, List<String>> anagrams = new HashMap<>();
   //Go through each string in the input array and check if the string is anagram of any keys in the 
   //hash map. Add string to key. Otherwise, add current string's sorted string and current string
   //as a new key,value pair to the hash map
    for(String str : strs) {
      char[] sortedStr = str.toCharArray();
      Arrays.sort(sortedStr);
      String key = String.valueOf(sortedStr);
      if (anagrams.containsKey(key) == false) {
        List<String> newsubList = new ArrayList<>();
        anagrams.put(key, newsubList);
      }
      anagrams.get(key).add(str);
    }
    for (Map.Entry<String, List<String>> anagramList : anagrams.entrySet()) {
      output.add(anagramList.getValue());
    }

    return output;
  }
} 
